### PR TITLE
Proof of concept: Use a trait for parsers

### PR DESCRIPTION
**Pull request by gmponos**
Made all the parsers use the same code by using a trait..

At my initial commits I made all the parsers use the same code. And then later I replaced the code with a trait.

@frederikbosch made a comment [here](https://github.com/moneyphp/money/pull/520#issuecomment-444468991)

I had also thought about the same idea but later on my plan was to submit a PR just like this one.

Marking this trait as internal will also not be inside the BC promise.

### CODE REVIEW

**Comment by Big-Shark** [comment]
@gmponos Any updates?

**Comment by gmponos** [comment]
This could be my fault.. but I need some feedback in order to proceed and finalize this or not.

Neither on the current PR or [here](https://github.com/moneyphp/money/pull/520#issuecomment-444468991) I got any feedback.

**Comment by Big-Shark** [comment]
@gmponos This is look very nice for me, but I am not scure, why we can not use `number_format` ?
Yasterday I created this class https://gist.github.com/Big-Shark/84434409e4f4cb6754645ea283e48c74
And all my tests passed.
Can you explorer this?

File: src/Parser/DecimalParserTrait.php [hunk 1]
```diff
@@ -0,0 +1,62 @@
+<?php
+
+namespace Money\Parser;
+
+use Money\Currency;
+use Money\Exception\ParserException;
+use Money\Money;
+use Money\Number;
+
+trait DecimalParserTrait
```
Comment (sagikazarmark) [inline src/Parser/DecimalParserTrait.php]:
```suggestion
/**
 * Common decimal parsing logic extracted from DecimalMoneyParser for reuse.
 */
trait DecimalParser
```

File: src/Parser/DecimalParserTrait.php [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+{
+    public static $decimalPattern = '/^(?P<sign>-)?(?P<digits>0|[1-9]\d*)?\.?(?P<fraction>\d+)?$/';
```
Comment (sagikazarmark) [inline src/Parser/DecimalParserTrait.php]: This should be a const IMHO

File: src/Parser/DecimalMoneyParser.php [hunk 3]
```diff
@@ -16,7 +14,7 @@
  */
 final class DecimalMoneyParser implements MoneyParser
 {
-    const DECIMAL_PATTERN = '/^(?P<sign>-)?(?P<digits>0|[1-9]\d*)?\.?(?P<fraction>\d+)?$/';
```
Comment (sagikazarmark) [inline src/Parser/DecimalMoneyParser.php]: Note that removing this constant is probably a BC break.

**Comment by sagikazarmark** [comment]
Like the idea! Can you please rebase the branch from master?

File: src/Parser/DecimalParserTrait.php (hunk 2 above)
Comment (gmponos) [reply src/Parser/DecimalParserTrait.php]: Traits can not have constants :)

File: src/Parser/DecimalMoneyParser.php (hunk 3 above)
Comment (gmponos) [reply src/Parser/DecimalMoneyParser.php]: reverted

File: src/Parser/DecimalParserTrait.php (hunk 2 above)
Comment (sagikazarmark) [reply src/Parser/DecimalParserTrait.php]: In that case, I would make it protected.

Comment (gmponos) [reply src/Parser/DecimalParserTrait.php]: Done.

**Comment by gmponos** [comment]
Marked as WIP.. let's merge this first: https://github.com/moneyphp/money/pull/586

**Comment by frederikbosch** [comment]
Sorry @gmponos, I think we are going to leave this as is. In my opinion traits are to be avoided.

**Pull request closed without merging**